homeassistant/components/sensor/geo_rss_events.py

```diff
@@ -0,0 +1,359 @@
+        self._unit_of_measurement = unit_of_measurement
+        if category is not None:
+            id_base = category
+        else:
+            id_base = 'any'
+        if name is not None:
+            id_base = '{}_{}'.format(name, id_base)
+        self.entity_id = async_generate_entity_id(ENTITY_ID_FORMAT, id_base,
+                                                  hass=hass)
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        if self._category is not None:
+            return self._category
+        else:
+            return 'Any'
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        if isinstance(self._state, list):
+            return len(self._state)
+        else:
+            return self._state
+
+    @property
+    def should_poll(self):  # pylint: disable=no-self-use
+        """No polling needed."""
+        return False
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement."""
+        return self._unit_of_measurement
+
+    @property
+    def icon(self):
+        """Return the icon to use in the frontend, if any."""
+        return self._icon
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        matrix = {}
+        if self._state is not STATE_UNKNOWN:
+            for event in self._state:
+                matrix[event.title] = '{:.0f}km'.format(event.distance)
+        return matrix
+
+    def update(self):
+        """Update this sensor from the GeoRSS service."""
+        all_events = self._data.events
+        if self._category is None:
+            # Add all events regardless of category.
+            self._state = all_events
+        else:
+            # Group events by category.
+            my_events = []
+            for event in all_events:
+                if event.category == self._category:
+                    my_events.append(event)
+            self._state = my_events
+            _LOGGER.info("New state: %s", self._state)
+
+
+class GeoRssServiceData(object):
+    """Stores the latest data from the GeoRSS service."""
+
+    def __init__(self):
+        """Initialize the data object."""
+        self.events = None
+
+
+class GeoRssServiceUpdater:
+    """Provides access to GeoRSS feed and creates and updates UI devices."""
+
+    def __init__(self, hass, data, home_latitude, home_longitude, url,
+                 radius_in_km, devices):
+        """Initialize the update service."""
+        self._hass = hass
+        self._data = data
+        self._home_coordinates = [home_latitude, home_longitude]
+        self._url = url
+        self._radius_in_km = radius_in_km
+        self._devices = devices
+
+    @asyncio.coroutine
+    def async_update(self, *_):  # pragma: no cover
+        """Retrieve data from GeoRSS feed and update devices."""
+        import feedparser
+        feed_data = feedparser.parse(self._url)
```

Comment:
you could also fetch the url asynchronously  and process the result locally with feedparser.